raster-foundry/granary
# Setup CI with CircleCI and Quay

## Overview

- Add `.circleci/config.yml` file
- Add README badges 😎
- Use Scala 2.12 artifacts in `granary-api` container image
- Fix `scapegoatVersion` scope
- Add `app-backend/database/src/main/resources/db/.gitkeep` so `granary-api-migrations` can build
- Use Paul Phillip's more capable [sbt runner](https://github.com/paulp/sbt-extras)
- Add `shellcheck` Docker Compose service
- Add `cipublish` to publish container images to [Quay.io](https://quay.io/organization/raster-foundry)
- Add mechanism for generating `.sbtopts` to resolve CI build failures
- Align impacted STRTA with the [ADR](https://github.com/azavea/architecture/blob/master/doc/arch/adr-0000-scripts-to-rule-them-all.md) and Google's [Shell Style Guide](https://google.github.io/styleguide/shell.xml)

Resolves #8 

### Notes

I wanted to get Slack notifications wired up. It seems there are two paths we could go down. CircleCI lets you [configure notifications](https://circleci.com/docs/2.0/notifications/#enable-chat-notifications) per project, but there's no way to filter based on branch, so this would be information overload. There's also the [Slack orb](https://github.com/CircleCI-Public/slack-orb). This option seemed equally inflexible. We wouldn't be able to gate against PRs or have notification for fixes. Also, it requires using the 2.1 config version, which seems to break things. From what I can tell, it's [still in alpha](https://discuss.circleci.com/t/2-1-config-and-build-processing/24102/11).

Also, I wanted to split out `cipublish` execution into a `deploy` job, but there's no way to share the Docker execution environment across jobs. Here's a related [thread](https://discuss.circleci.com/t/can-docker-images-be-preserved-between-jobs-in-a-workflow-without-a-manual-load-save/23388/11).

## Testing Instructions

- See CircleCI build: https://circleci.com/gh/raster-foundry/granary/21
- See container images on Quay.io: https://quay.io/repository/raster-foundry/granary-api
